### ISSUE TITLE
can't compare offset-naive and offset-aware datetimes

### ISSUE THREAD

**Issue by switchtrue** [stevearc/pypicloud#266]
I'm using pypicloud with gcs storage and redis cache. When clicking "Rebuild package list" in the admin UI I'm getting `TypeError: can't compare offset-naive and offset-aware datetimes` (stacktrace below).

It's being raised by [this line of code](https://github.com/stevearc/pypicloud/blob/master/pypicloud/cache/redis_cache.py#L269) but I can't quite figure out what that's trying to achieve, nor how it's ended up in this situation.

```
ERROR 2020-11-12 08:12:33,355 [pypicloud.views] can't compare offset-naive and offset-aware datetimes
Traceback (most recent call last):
  File "/opt/pypicloud/venv/lib/python3.8/site-packages/pyramid/tweens.py", line 41, in excview_tween
    response = handler(request)
  File "/opt/pypicloud/venv/lib/python3.8/site-packages/pyramid/router.py", line 147, in handle_request
    response = _call_view(
  File "/opt/pypicloud/venv/lib/python3.8/site-packages/pyramid/view.py", line 683, in _call_view
    response = view_callable(context, request)
  File "/opt/pypicloud/venv/lib/python3.8/site-packages/pyramid/config/views.py", line 188, in attr_view
    return view(context, request)
  File "/opt/pypicloud/venv/lib/python3.8/site-packages/pyramid/config/views.py", line 214, in predicate_wrapper
    return view(context, request)
  File "/opt/pypicloud/venv/lib/python3.8/site-packages/pyramid/viewderivers.py", line 325, in secured_view
    return view(context, request)
  File "/opt/pypicloud/venv/lib/python3.8/site-packages/pyramid/viewderivers.py", line 436, in rendered_view
    result = view(context, request)
  File "/opt/pypicloud/venv/lib/python3.8/site-packages/pyramid/viewderivers.py", line 116, in _class_requestonly_view
    response = getattr(inst, attr)()
  File "/opt/pypicloud/venv/lib/python3.8/site-packages/pypicloud/views/admin.py", line 26, in rebuild_package_list
    self.request.db.reload_from_storage()
  File "/opt/pypicloud/venv/lib/python3.8/site-packages/pypicloud/cache/redis_cache.py", line 269, in reload_from_storage
    if package.last_modified > summary["last_modified"]:
TypeError: can't compare offset-naive and offset-aware datetimes
```

**Comment by switchtrue** [comment]
I've added some additional debug in an exception handler.

Additional logging:
```
LOG.error(package.last_modified)
LOG.error(summary["last_modified"])
LOG.error(type(package.last_modified))
LOG.error(type(summary["last_modified"]))
```

Result:
```
ERROR 2020-11-12 08:31:45,643 [pypicloud.cache.redis_cache] 2020-10-21 13:28:43.480000+00:00
ERROR 2020-11-12 08:31:45,643 [pypicloud.cache.redis_cache] 2020-10-21 13:28:43.480000
ERROR 2020-11-12 08:31:45,643 [pypicloud.cache.redis_cache] <class 'datetime.datetime'>
ERROR 2020-11-12 08:31:45,643 [pypicloud.cache.redis_cache] <class 'datetime.datetime'>
```

It seems that [_load_summaries](https://github.com/stevearc/pypicloud/blob/d74cdd710009ce8a343b9cf1b83a413ae925dfd4/pypicloud/cache/redis_cache.py#L110) of the redis cache returns `last_modified` as `datetime.utcfromtimestamp` which is [naive](https://docs.python.org/3/library/datetime.html#datetime.datetime.utcfromtimestamp).

But also, google cloud storage is [returning a naive datetime](https://github.com/googleapis/python-storage/blob/master/google/cloud/storage/blob.py#L1153). I can't see how the offset is being introduced.

**Comment by stevearc** [comment]
It looks like GCS is actually returning a timezone-aware datetime. The function you linked is checking [self.updated](https://github.com/googleapis/python-storage/blob/142639a9b9621b17709ee631652a6ef4f5058caf/google/cloud/storage/blob.py#L3526-L3539), which is in turn calling `_rfc3339_to_datetime()`, and that is [attaching a timezone](
https://github.com/googleapis/python-cloud-core/blob/58c88db4e5c175d52ab796c519dd4a4ed2ec106e/google/cloud/_helpers.py#L277-L286).

There have already been some other issues from this because the DynamoDB backend uses tz-aware datetimes as well. I think the fix is to just *always* use tz-aware datetimes in all locations. I've pushed up a branch. Can you give this version a try and let me know if it works for you?
https://github.com/stevearc/pypicloud/tree/stevearc-tz-datetime

**Comment by switchtrue** [comment]
That works! Thank you.

Playing around with though has lead me to what I believe is another issue. If I have `db.graceful_reload = false` this original issue never occurred, I guess there was no need to compare the date times as it just reloaded everything? However, swapping back to `db.graceful_reload = true` to correctly test this branch succeeds but there are a lot of packages missing. If I turn off `graceful_reload` all the packages are there. This occurs even if I flush the database and rebuild from scratch. This also occurs in both this branch and in version `1.1.6` that I was previously using.

I've attached screenshots of an example package - the only difference with these was changing `db.graceful_reload` and then rebuilding the cache.

# graceful_reload = true
![graceful_reload_true](https://user-images.githubusercontent.com/1535972/99031257-c85b9a00-25ca-11eb-9c5e-28218463b0af.png)

# graceful_reload = false
![graceful_reload_false](https://user-images.githubusercontent.com/1535972/99031251-c560a980-25ca-11eb-9192-31807343a517.png)

Do you want me to raise another issue for this?

**Comment by stevearc** [comment]
Great! I'll merge that change.

The problem with graceful_reload is a second issue, but I think I know what's happening. Open another issue for it and I'll see if I can get a fix up for it soon

**Comment by stevearc** [comment]
You can check out the fix for the graceful reload issue at #267
